### ISSUE TITLE
Add Database export and import as json arrays.

### ISSUE THREAD

**Issue by maxpiva** [japanesemediamanager/jmmserver#370]
That will enable flawless database migration between database types.

**Comment by bigretromike** [comment]
Your cache read everything right? Why not read database to cache and export cache ? that way you wouldn't have to figure out syntax and etc ;-)

But you probably already started it

**Comment by maxpiva** [comment]
not everything, in the last commit, you'll see what repositories are cached and whatnot, i think 1/3 of the repositories are cached the other not, anyway, since IRepository is the base interface, to dump, you only need to do GetAll(), and then store as json file.

**Comment by da3dsoul** [comment]
Is this in progress? We should remove it from 3.7 if not.

**Comment by da3dsoul** [comment]
In response to https://github.com/ShokoAnime/ShokoServer/issues/693, this should have the ability to filter what to export and import. The rules will need to be set, as there are ID references and such to consider.
For example:
- AniDB Data (in case you want guaranteed fresh data)
- Files (VideoLocals and places, this will need to store with a map of import folders and give the ability to remap on import)
- File states (watched states, MyList, etc. A combination of VideoLocal_User and settings)
- Links (Manual Links, ignored status, variation for multiple files, and CrossRef_File_Episode)
- Community Links (in case you manually linked them and they weren't approved yet)
- Groups (This will be stored with AniDB series IDs to make sure they match)
- Series Data (this will be AnimeSeries, along with overridden names and such)
- Group Filters
- Community site data (in case you want fresh data)
- Ignored Anime for recommendations
- Votes (in case you have it set not to sync, you can back up local votes)

That's all I can think of right now, but it's a start.

**Comment by da3dsoul** [comment]
I don't know if it needs to be JSON, as JSON is slow, but it'll need serialized one way or another. Protobuf or another binary serialization method could work as well.

**Comment by Cazzar** [comment]
JSON is slow because of reflection, and boxing, there isn't a lot that can be done, also in this case, would it be necessary for this to be a lightning quick procedure?

**Comment by da3dsoul** [comment]
No, but it shouldn't take all day either

**Comment by Cazzar** [comment]
You're imposing assumptions before they are even tested.

**Comment by da3dsoul** [comment]
Yes but my assumption is that a reflection operation on every model in the database is going to be slow. If I'm wrong, then thank God, but I doubt that I am.

**Comment by Cazzar** [comment]
Why would we need to export every model? We could just export core data
like links. And then re-populate the rest?

Also potentially have the option for a full dump?

**Comment by maxpiva** [comment]
Serializing al to json should be fast enough, the whole idea is not over
complex the whole thing, just touch, select a directory, and done. Every
table will get a [table].json with the all the data in there. i don't
expect to be more than few minutes on 3000+ series. The good thing about
this, is sql agnostic, so you could use to have a backup, or translate your
whole collection to other database. The only thing we should be aware, is
the import order, or temporary remove foreign references, autonumbers, etc
before import.
We could add filters after, but the base functionality should be dump the
whole db.
